src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1175,6 +1150,149 @@ internal bool HasCollectionExpressionApplicableAddMethod(SyntaxNode syntax, Type
+                            bool wasFullyInferred = false;
+
+                            if (thisTypeIsOpen)
+                            {
+                                constructed = ReducedExtensionMethodSymbol.InferExtensionMethodTypeArguments(
+                                                            candidate.Member, receiverType, addMethodBinder.Compilation, ref useSiteInfo, out wasFullyInferred);
+                            }
+
+                            if (constructed is null || !wasFullyInferred)
+                            {
+                                // It is quite possible that inference failed because we didn't supply type from the second argument
+                                if (typeParameters.Any((typeParameter, parameter) => parameter.Type.ContainsTypeParameter(typeParameter), candidate.Member.Parameters[1]))
+                                {
+                                    // Let's attempt inference with type for the second parameter
+                                    // We are going to use the second parameter's type for that
+                                    var definition = candidate.Member;
+                                    var argumentRefKinds = ArrayBuilder<RefKind>.GetInstance(2, RefKind.None);
+
+                                    OverloadResolution.GetEffectiveParameterTypes(
+                                        definition,
+                                        argumentCount: 2,
+                                        argToParamMap: default,
+                                        argumentRefKinds: argumentRefKinds,
+                                        isMethodGroupConversion: false,
+                                        allowRefOmittedArguments: methodGroup.ReceiverOpt.IsExpressionOfComImportType(),
+                                        binder: addMethodBinder,
+                                        expanded: candidate.Result.Kind == MemberResolutionKind.ApplicableInExpandedForm,
+                                        parameterTypes: out ImmutableArray<TypeWithAnnotations> parameterTypes,
+                                        parameterRefKinds: out ImmutableArray<RefKind> parameterRefKinds);
+
+                                    argumentRefKinds.Free();
+
+                                    TypeSymbol secondArgumentType = candidate.Member.Parameters[1].Type;
+
+                                    if (constructed?.TypeSubstitution is { } typeMap)
+                                    {
+                                        // If we were able to infer something just from the first parameter,
+                                        // apply that to the second type, otherwise inference is going to fail
+                                        // for those type parameters.
+                                        secondArgumentType = typeMap.SubstituteType(secondArgumentType).Type;
+                                    }
```

Comment:
Can we take the type from the constructed parameter directly?
```
TypeSymbol secondArgumentType = (constructed ?? candidate.Member).Parameters[1].Type;
```

Reply:
> Can we take the type from the constructed parameter directly?

It looks like this works too. Good idea. Makes things simpler